doc/ref/modules/index.rst

```diff
@@ -376,6 +376,22 @@ The above example will force the minion to use the :py:mod:`systemd
 
 .. __: https://github.com/saltstack/salt/issues/new
 
+Logging Restrictions
+--------------------
+
+As a rule, logging should not be done anywhere in a file before a module is
+loaded. This rule apples to all code that would run before the ``__virtual__()``
+function, as well as the code within the ``__virtual__()`` function itself.
+
+If logging statements are called before the ``__virtual__()`` function determines
+the module should be loaded, then those logging statements will be called
+repeatedly. This clutters up log files unnecessarily.
+
+Exceptions may be considered for logging statements made at the ``trace`` level.
+However, it is better to provide the necessary information by another means.
+One method is to :ref:`return error information <modules-error-info>` in the
+``__virtual__()`` function.
```

Comment:
Good, but I think we need to then link to the part of the docs where that is explained. Otherwise, show a brief example here.

Reply:
@cachedout I fixed the other two comments, but this one is already linked to the docs with the `ref` tag.